File: go/src/sample/mocks/smokedetector.go
package mocks

import (
	"time"
	"v.io/v23/rpc"
)

const (
	// Smoke detector status constants
	smokeDetectorDetecting    = "smoky"
	smokeDetectorNotDetecting = "normal"

	smokeDetectorDefaultSensitivity = int16(10)
)

// SmokeDetector allows clients to monitor and adjust a smoke detector.
type smokeDetector struct {
	status      string
	sensitivity int16
}

// Status retrieves the current status and sensitivity of the SmokeDetector.
func (s *smokeDetector) Status(rpc.ServerCall) (status string, sensitivity int16, err error) {
	return s.status, s.sensitivity, nil
}

// Test the SmokeDetector to check if it is working.
func (s *smokeDetector) Test(rpc.ServerCall) (bool, error) {
	time.Sleep(1500 * time.Millisecond) // simulate testing for 1.5 seconds
	success := s.sensitivity > 0        // succeed only if sensitivity is positive
	return success, nil
}

// Sensitivity adjusts the SmokeDetector's sensitivity to smoke.
func (s *smokeDetector) Sensitivity(_ rpc.ServerCall, sensitivity int16) error {
	s.sensitivity = sensitivity
	return nil
}

// NewSmokeDetector creates a new smoke detector stub.
func NewSmokeDetector() *smokeDetector {
	return &smokeDetector{
		status:      smokeDetectorNotDetecting,
		sensitivity: smokeDetectorDefaultSensitivity,
	}
}

File: src/services/sample-world/smokedetector.js
// Copyright 2015 The Vanadium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

var vdl = require('./ifc');

module.exports = SmokeDetector;

var DETECTING = 'smoky'; // jshint ignore:line
var NOT_DETECTING = 'normal';
var DEFAULT_SENSITIVITY = 10;
var TEST_DURATION = 1500;

// SmokeDetector allows clients to monitor and adjust a smoke detector.
function SmokeDetector() {
  this.state = NOT_DETECTING;
  this.sense = DEFAULT_SENSITIVITY;
  this.handle = null;
}

SmokeDetector.prototype = new vdl.SmokeDetector();

// Status retrieves the current status and sensitivity of the SmokeDetector.
SmokeDetector.prototype.status = function(context, serverCall) {
  return [this.state, this.sense];
};

// Test the SmokeDetector to check if it is working.
SmokeDetector.prototype.test = function(context, serverCall, cb) {
  var self = this;

  // Wait until the TEST_DURATION is over.
  // Then succeed only if sensitivity is positive.
  setTimeout(function() {
    cb(null, self.sense > 0);
  }, TEST_DURATION);
};

// Sensitivity adjusts the SmokeDetector's sensitivity to smoke.
SmokeDetector.prototype.sensitivity =
  function(context, serverCall, sensitivity) {

  this.sense = sensitivity;
};